tests/index.test.ts

```diff
@@ -358,21 +370,50 @@ describe.each([
 describe.each([
   ['fetchFetcher & ethersContractCaller', fetchFetcher, ethersContractCaller],
   ['axiosFetcher & web3ContractCaller', axiosFetcher, web3ContractCaller],
+  [
+    'axiosFetcher & customGanacheContractCaller',
+    axiosFetcher,
+    customGanacheContractCaller,
+  ],
 ])(
   'ParaSwap SDK: contract calling methods: %s',
   (testName, fetcher, contractCaller) => {
-    let paraSwap: ApproveTokenFunctions<any> & GetSpenderFunctions;
+    type ApproveTxResult =
+      | ethers.ContractTransaction
+      | Web3UnpromiEvent
+      | string;
+    // @TODO try Instantiation Expression when TS 4.7 `as constructApproveToken<TxResponse>`
+    type ApproveConstructor = (
+      options: ConstructProviderFetchInput<ApproveTxResult>
+    ) => ApproveTokenFunctions<ApproveTxResult>;
+
+    let paraSwap: ApproveTokenFunctions<ApproveTxResult> & GetSpenderFunctions;
 
     beforeAll(() => {
-      paraSwap = constructPartialSDK(
+      paraSwap = constructPartialSDK<
+        SDKConfig<ApproveTxResult>,
+        [ApproveConstructor, typeof constructGetSpender]
+      >(
         { network, fetcher, contractCaller },
         constructApproveToken,
         constructGetSpender
       );
     });
     test('approveToken', async () => {
       const tx = await paraSwap.approveToken('12345', DAI);
-      await tx.wait(1);
+
+      if (typeof tx === 'string') {
+        // tx is sent
+        // wait for it to mine however you prefer
+        await ethersProvider.waitForTransaction(tx);
+      } else if ('wait' in tx) {
+        await tx.wait(1);
+      } else if ('on' in tx) {
+        await new Promise<Web3TransactionReceipt>((resolve, reject) => {
+          tx.on('receipt', resolve);
```

Comment:
is it a bad idea to unify it in the SDK so that would always return `string`?

Reply:
SimpleSDK in one of the later PRs does exactly that